### ISSUE TITLE
Internal compiler error when attempting to implement an associated const with a local variable

### ISSUE THREAD

**Issue by petertodd** [rust-lang/rust#44239]
Specifically this code:

```
trait Foo {
    const N: usize;
}

fn main() {
    struct MyFoo;

    let n = 0;

    impl Foo for MyFoo {
        const N: usize = n;
    }   
}
```

results in the following error message:

```
$ rustc test.rs 
error: internal compiler error: /checkout/src/librustc_typeck/check/mod.rs:1846: no type for local variable local n (id=15)
  --> test.rs:11:26
   |
11 |         const N: usize = n;
   |                            ^

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.21.0-nightly (97b01abf3 2017-08-31) running on x86_64-unknown-linux-gnu

thread 'rustc' panicked at 'Box<Any>', /checkout/src/librustc_errors/lib.rs:439:8
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

Expected behavior: While I can understand if this level of meta isn't possible, it should give a proper error message! :)

**Comment by seventh-chord** [comment]
Just to note, a slightly simpler case which triggers the same error is to have the associated const in a impl block for the struct directly:

```
fn main() {
    let n = 0;

    struct Foo;
    impl Foo {
        const N: usize = n;
    }   
}
```

Error message:

```
error: internal compiler error: /checkout/src/librustc_typeck/check/mod.rs:1785: no type for local variable local n (id=8)
 --> test.rs:6:26
  |
6 |         const N: usize = n;
  |                          ^

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.20.0 (f3d6973f4 2017-08-27) running on x86_64-unknown-linux-gnu

note: run with `RUST_BACKTRACE=1` for a backtrace

thread 'rustc' panicked at 'Box<Any>', /checkout/src/librustc_errors/lib.rs:437:8
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace: (I have cut out the stacktrace for brevity)
```
